### DISCUSSION TITLE
Securing JWT

### DISCUSSION THREAD

**Discussion by owen2jan** [nuxt/nuxt#28498] [Questions]
I am looking for a secure way of authenticating with jwt tokens. My initial idea was to store it in a http only cookie with samesite lax. I am looking for any examples of how this can be done "properly" and any advice on security concerns that I should have. 

Also, since I am using an API backend, what are the advantages of having my request in the /server/api directory instead of just having, for example, a useFetch in my /pages/products.vue .

**Comment by YuliyaMinsk** [comment]
I've recently been working on a similar setup using JWT tokens, and I'd like to share my approach with you. I'm also still in the process of refining it, so I'd appreciate any feedback or insights, especially regarding how you've handled the refresh token mechanism.

**My Approach**

I decided to store the JWT tokens in HTTP-only cookies with SameSite=Lax for added security. To ensure the tokens are secure, I created a server-side API endpoint in server/api/auth/set-tokens.ts, where I handle the token encryption and set the cookies. Here's a simplified version of my implementation:

```ts
export default defineEventHandler(async (event) => {
  const { access_token, refresh_token } = await readBody(event);
  const isProduction = process.env.NODE_ENV === 'production';

  const encryptedAccessToken = encryptValue(access_token);
  const encryptedRefreshToken = encryptValue(refresh_token);

  setCookie(event, 'access_token', encryptedAccessToken, {
    httpOnly: true,
    secure: isProduction,
    sameSite: 'lax',
    path: '/',
  });

  setCookie(event, 'refresh_token', encryptedRefreshToken, {
    httpOnly: true,
    secure: isProduction,
    sameSite: 'lax',
    path: '/',
    maxAge: 3600, // 1 hour
  });

  return { success: true };
});
```

I opted for HTTP-only cookies because they are inaccessible to JavaScript, reducing the risk of XSS attacks. This setup, combined with SameSite=Lax, helps protect against CSRF attacks.

I'm currently working on the refresh token mechanism, and I'm curious about your approach. How have you organized the token refresh process in your application? Specifically, how do you handle the scenario where the access token expires and needs to be refreshed using the refresh token?

Any insights you can share would be greatly appreciated!
